### ISSUE TITLE
implement ARAX_resultify.py

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#618]

**Comment by saramsey** [comment]
ok @edeutsch and @dkoslicki I think it is working. please let me know if this is what you were looking for

**Comment by saramsey** [comment]
have checked in 12 unit tests, they all are working

**Comment by dkoslicki** [comment]
Btw, is anyone planning on writing the wrapper to get this to integrate with the `demo` branch? In emails, @edeutsch may have indicated that he would work on this (possibly the week after this Monday). LMK if that’s the plan, otherwise we need to identify who can do that (as they need to be familiar with the API object model).

**Comment by saramsey** [comment]
I'm working on a new version of this module that (1) deals with the is_set=True case and (2) complies with the API object model.

**Comment by saramsey** [comment]
I have attempted to write a function to generate a Results list given a KG and a QG (all three in the Swagger API object model). The code is in the "demo" branch:

https://github.com/RTXteam/RTX/blob/demo/code/ARAX/kg_qg_results.py

**Comment by dkoslicki** [comment]
@saramsey Not all nodes in the KG need to have the `is_set` property specified so [these lines](https://github.com/RTXteam/RTX/blob/demo/code/ARAX/kg_qg_results.py#L126-L129) shouldn't return an error. Eg.
```python
{'description': 'A fluid filled cavity that develops with a bone.',
  'id': 'HP:0012062',
  'name': 'Bone cyst',
  'node_attributes': None,
  'qnode_id': 'n01',
  'symbol': None,
  'type': ['phenotypic_feature'],
  'uri': 'http://purl.obolibrary.org/obo/HP_0012062'}
```
is an acceptable node in the KG corresponding to the QG:
```
{'edges': [{'id': 'e00',
            'negated': None,
            'relation': None,
            'source_id': 'n00',
            'target_id': 'n01',
            'type': 'has_phenotype'}],
 'nodes': [{'curie': 'DOID:8398',
            'id': 'n00',
            'is_set': None,
            'type': 'disease'},
           {'curie': None,
            'id': 'n01',
            'is_set': True,
            'type': 'phenotypic_feature'}]}
```
Better to assume that if `node.is_set` is `None` (or `False`), then it's not supposed to map to a set of QG nodex, but a single QG node specified by the `qnode_id`.

**Comment by dkoslicki** [comment]
Looks like another problem with the consistency of edge-to-node relationships (starting on line 143 of `kg_qg_results.py`:
```python
results = R.get_results_for_kg_by_qg(message.knowledge_graph, message.query_graph)
Traceback (most recent call last):
  File "/home/dkoslicki/miniconda3/envs/ARAX/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3331, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-45-d718b03fa131>", line 1, in <module>
    results = R.get_results_for_kg_by_qg(message.knowledge_graph, message.query_graph)
  File "/home/dkoslicki/Desktop/RTX/code/ARAX/kg_qg_results.py", line 154, in get_results_for_kg_by_qg
    kg_target_node_id + ", hav no corresponding edge in the QG")
ValueError: The two nodes for KG edge 1007224, UniProtKB:P08173 and CHEMBL.COMPOUND:CHEMBL697, hav no corresponding edge in the QG
```
yet looking at this edge:
```python
[x for x in list(message.knowledge_graph.edges) if x.id=='1007224']
Out[46]: 
[{'confidence': 0.16580056931,
  'defined_datetime': None,
  'edge_attributes': None,
  'evidence_type': None,
  'id': '1007224',
  'is_defined_by': 'RTX',
  'negated': None,
  'provided_by': 'ChEMBL',
  'publications': None,
  'qedge_id': 'e02',
  'qualifiers': None,
  'relation': 'targets',
  'source_id': 'UniProtKB:P08173',
  'target_id': 'CHEMBL.COMPOUND:CHEMBL697',
  'type': 'physically_interacts_with',
  'weight': None}]
```
its `qedge_id` is 'e02' and in the query graph
```python
message.query_graph.edges
Out[47]: 
[{'id': 'e00',
  'negated': None,
  'relation': None,
  'source_id': 'n00',
  'target_id': 'n01',
  'type': 'has_phenotype'},
 {'id': 'e01',
  'negated': None,
  'relation': None,
  'source_id': 'n02',
  'target_id': 'n01',
  'type': 'indicated_for'},
 {'id': 'e02',
  'negated': None,
  'relation': None,
  'source_id': 'n02',
  'target_id': 'n03',
  'type': None},
 {'id': 'J1',
  'negated': None,
  'relation': 'jaccard_index',
  'source_id': 'n00',
  'target_id': 'n02',
  'type': 'J1'}]
```
There's definitely an 'e02' and the nodes even match up with the right `source_id` and `target_id`:
```python
 message.query_graph.nodes
Out[48]: 
[{'curie': 'DOID:8712', 'id': 'n00', 'is_set': None, 'type': 'disease'},
 {'curie': None, 'id': 'n01', 'is_set': True, 'type': 'phenotypic_feature'},
 {'curie': None, 'id': 'n02', 'is_set': True, 'type': 'chemical_substance'},
 {'curie': None, 'id': 'n03', 'is_set': True, 'type': 'protein'}]
```

**Comment by dkoslicki** [comment]
**Sub-issue label: # 3**
We will want to address how we really want the results stored. After making hacky workarounds to sub-issues 1 and 2, I find that in our current demo example 2 (i.e. `ARAX_query.py 12`):
```python
query = { "previous_message_processing_plan": { "processing_actions": [
            "create_message",
            "add_qnode(name=DOID:14330, id=n00)",
            "add_qnode(type=protein, is_set=true, id=n01)",
            "add_qnode(type=chemical_substance, is_set=true, id=n02)",
            "add_qedge(source_id=n00, target_id=n01, id=e00)",
            "add_qedge(source_id=n01, target_id=n02, id=e01, type=physically_interacts_with)",
            "expand(edge_id=[e00,e01])",
            "overlay(action=compute_jaccard, start_node_id=n00, intermediate_node_id=n01, end_node_id=n02, virtual_edge_type=J1)",
            "filter_kg(action=remove_edges_by_attribute, edge_attribute=jaccard_index, direction=below, threshold=.2, remove_connected_nodes=t, qnode_id=n02)",
            "filter_kg(action=remove_edges_by_property, edge_property=provided_by, property_value=Pharos)",
            "return(message=true, store=true)",
            ] } }
```
Only a single result gets returned (basically, the entire KG). The goal of this example was to have the result "essence"'s be basically a list of drugs (eg `[x.name for x in message.knowledge_graph.nodes if x.type[0] == 'chemical_substance']`), and the results themselves showing how each of these connect back to parkinson's/DOID:14330. 

Maybe we need to expand the `add_qnode` syntax to indicate where each subgraph path should start/end. and then try to fulfill all QG paths from start to end (or _for each_ start  and _for each_ end if `is_set=True`). Thoughts @saramsey and @edeutsch?

**Comment by saramsey** [comment]
OK I am working on this -- I have been in class. Looking into these issues now....

**Comment by saramsey** [comment]
OK, just pushed a updated version to address sub-issue 1 (includes a test case, test02).

**Comment by saramsey** [comment]
have just pushed an updated version to address sub-issue 2 (includes a test case, test03)

**Comment by saramsey** [comment]
Hi @dkoslicki for sub-issue 3, can you please paste here a dump of the KG and QG that are being passed to `get_results_for_kg_by_qg`?

**Comment by saramsey** [comment]
sub-issue 4: add error handling so that exceptions are caught and wrapped in an API Message object

**Comment by saramsey** [comment]
OK @dkoslicki @edeutsch the module https://github.com/RTXteam/RTX/blob/demo/code/ARAX/kg_qg_results.py 
is ready for testing.

**Comment by saramsey** [comment]
Test case `test05` specifically tests sub-issue 4.

**Comment by dkoslicki** [comment]
@saramsey do you want me to hold off until it's placed into the `ARAXQuery` subfolder, wrapped in a class, and is tested within a `ARAX_query.py` DSL command? Or should I test before then? I can put it on the agenda for tomorrow either way.

**Comment by dkoslicki** [comment]
Ridiculously minor, but my IDE says line 57 `qg_nodes_override_treat_is_set_as_false: bool = True,`
and line 80 `qg_nodes_override_treat_is_set_as_false: set = None` have conflicting suggested types of `bool` or `set`. Don't think it'll matter much, but thought I'd mention it...

**Comment by saramsey** [comment]
From @dkoslicki 
```
Hi Steve,

I do concur with Eric re: message/responses and being in a class. Take a look at ARAX_query.py and ARAX_overlay.py. Your code should look generally like ARAX_overlay.py, and the optional parameters (like qg_nodes_override_treat_is_set_as_false) are passed when you call the DSL in ARAX_query.py (see towards the end of that file for examples.).

Thanks,

~David
```

**Comment by saramsey** [comment]
> Ridiculously minor, but my IDE says line 57 `qg_nodes_override_treat_is_set_as_false: bool = True,`
> and line 80 `qg_nodes_override_treat_is_set_as_false: set = None` have conflicting suggested types of `bool` or `set`. Don't think it'll matter much, but thought I'd mention it...

logging this as sub-issue 5

**Comment by saramsey** [comment]
sub-issue 5 fixed

**Comment by saramsey** [comment]
regarding sub-issue 6: minor note, I see that the suggested example module ARAX_query.py starts with
```
#!/bin/env python3
```
on L1, but my macOS doesn't have `/bin/env` and Ubuntu doesn't have it either. Seems like `/usr/bin/env` is the LCD here so that is what I will use, unless there is a reason not to.

**Comment by dkoslicki** [comment]
Re: `#!/bin/env python3`, that's left over from a copy-paste I did of one of Eric's `ARAX_query.py`
I don't use it, but imagine he might be (he's deving on Windows, so who knows), so left it in just in case.

**Comment by edeutsch** [comment]
It is fine to convert all to /usr/bin/env. I haven't been using them, I just put in that (erroneous) boilerplate. Also, I did not check them in a executable. So if we want the shebang to work on all, then we need to fix the shebang in all files and then also make them executable. I can do this in a few days, or someone else can beat me to it.

**Comment by dkoslicki** [comment]
@edeutsch Doesn't matter to me. I always `python ARAX_query.py` anyways and it's just for dev purposes. So I don't care about the shebangs or them being executable.

**Comment by dkoslicki** [comment]
**sub-issue 7**
`python ARAX_query.py 12` appears to be working
```python
query = { "previous_message_processing_plan": { "processing_actions": [
            "create_message",
            "add_qnode(name=DOID:14330, id=n00)",
            "add_qnode(type=protein, is_set=true, id=n01)",
            "add_qnode(type=chemical_substance, is_set=true, id=n02)",
            "add_qedge(source_id=n00, target_id=n01, id=e00)",
            "add_qedge(source_id=n01, target_id=n02, id=e01, type=physically_interacts_with)",
            "expand(edge_id=[e00,e01])",
            "overlay(action=compute_jaccard, start_node_id=n00, intermediate_node_id=n01, end_node_id=n02, virtual_edge_type=J1)",
            "filter_kg(action=remove_edges_by_attribute, edge_attribute=jaccard_index, direction=below, threshold=.2, remove_connected_nodes=t, qnode_id=n02)",
            "filter_kg(action=remove_edges_by_property, edge_property=provided_by, property_value=Pharos)",
            "resultify(ignore_edge_direction=true, qg_nodes_override_treat_is_set_as_false=[n02,n02])",
            "return(message=true, store=false)",
            ] } }
and is returning the expected number of different results.
```
**However**, `python ARAX_query.py 17` is not:
```
query = {"previous_message_processing_plan": {"processing_actions": [
            "create_message",
            "add_qnode(name=DOID:731, id=n00, type=disease, is_set=false)",
            "add_qnode(type=phenotypic_feature, is_set=false, id=n01)",
            "add_qedge(source_id=n00, target_id=n01, id=e00)",
            "expand(edge_id=e00)",
            'resultify(ignore_edge_direction=true)',
            "return(message=true, store=false)"]}}
```
Only a single result is being returned, when it should be a something like 3223 results as checked in cypher via:
```
match (n:disease{id:"DOID:731"})-[]-(m:phenotypic_feature) return count(distinct m)
```

**Comment by saramsey** [comment]
From @dkoslicki 

> They're quite minor things. Basically:
> `qg_nodes_override_treat_is_set_as_false` needs to be renamed to be something more understandable to an end user.
> We need a top level `describe_me()` method. An example of the one I put together for Amy's expander class is found here: https://github.com/RTXteam/RTX/blob/overlay/code/ARAX/ARAXQuery/ARAX_expander.py#L30-L51 Basically needs to return a list with a dict that has the keys brief_description and any other keys corresponding to optional parameters to `resultify()` whose values correspond to a set that gives human readable description of what they do (and if they are optional or required).

**Comment by saramsey** [comment]
Sub-issue 9 (reported by @dkoslicki ):

> Need to populate the other Result properties (such as essence, row_data, reasoner_id, result_code, etc.) as a number of these are essential for the UI

**Comment by edeutsch** [comment]
reasoner_id should get set during create_message():
https://github.com/RTXteam/RTX/blob/02fd462816a554eeebd820cc8e8e7dcf65ae3775/code/ARAX/ARAXQuery/ARAX_messenger.py#L53

so no need to do that, but I agree the rest should be set of something nice by the results generator.

**Comment by dkoslicki** [comment]
@saramsey see #642 re: the problem I encountered last night (this morning?) wrt storing the messages. Looks like a single line fix, thankfully

**Comment by saramsey** [comment]
@dkoslicki suggests rephrasing along these lines:
> “”“`resultify` creates a list of results from the input query graph (QG) based on the the information contained in the message knowledge graph (KG). Every subgraph through the KG that satisfies the GQ is returned. Such use cases include:
> \n \n
> * `resultify()` returns all subgraph in the knowledge graph that satisfy the query graph.
> * `resultify(force_isset_false=n01)` this forces each result to include only one example of node `n01` if it was originally part of a set in the QG. This can be used to, for example, <insert example re: one of the demo examples>
> * `resultify(ignore_edge_direction=True)` this ignores edge directions in the QG in case the underlying KG has directional edges. For example, you may want to include results that include relationships like `(protein)-[involved_in]->(pathway)` even thought the underlying KG only contains directional edges of the form `(protein)<-[involved_in]-(pathway)` # I’m sure a better example can be come up with, but you know, it’s 1:30am over here
> 
> Note that this command will successfully execute given an arbitrary query graph and knowledge graph provided by the automated reasoning system, not just ones generated by team ARA Expander.
> “””
> 
> Then in the actual __DSL arguments__ you can list just the bare necessities of what is required such as:
> Force_isset_false: {‘a node ID contained in the QG’}
> Ignore_edge_direction: {‘True’, ‘False’}
> 
> This is intended to keep the document:
> 1. Readable for the NCATS review team (in particular, the ‘brief_description’ part)
> 2. Easily parsed by the RTX team in viewing the __DSL arguments__ to know that kinds of things they can pass in the commands.
> 
> I.e. broad descriptions of parameters go in ‘brief_descriptions’, and __DSL parameters__ is intended to be very brief and meant mainly for us and knowledgeable users who want to use the API.
> 
> LMN if that makes any sense or not.
>

**Comment by saramsey** [comment]
> And just FYI, you don’t need to follow the “action=“ paradigm that Finn and I have been using for ARAX_overlay.py and ARAX_filter_kg.py. So no need to provide an action (see Amy’s Expander code for the example). 

@dkoslicki  your script `document_dsl_commands.py` (version in the `demo` branch) on L58 specifically references an `action` key in the dictionary returned from `describe_me()`. If I do not set the `action` key as I am doing on L78 in `ARAX_resultify.py`, the `document_dsl_commands.py` script gives a `KeyError`.

**Comment by saramsey** [comment]
sub-issue 11 (from @dkoslicki ):

> And just FYI, you don’t need to follow the “action=“ paradigm that Finn and I have been using for ARAX_overlay.py and ARAX_filter_kg.py. So no need to provide an action (see Amy’s Expander code for the example). If you need assistance with prettifying it, LMN.

**Comment by saramsey** [comment]
sub-issue 12: use double-underscore nomenclature for module functions that are not intended to be public

**Comment by saramsey** [comment]
OK @dkoslicki I think I have completed the prettifying and refactoring of `ARAX_resultify.py`. All tests pass, including running `document_dsl_commands.py` which processes `ARAX_resultify.py`. The resulting markdown file `DSL_Documentation.md` looks reasonable (I think). Ready for your inspection.

**Comment by saramsey** [comment]
I propose to close this issue if there are no further sub-issues. What do you think, @edeutsch and @dkoslicki ?

**Comment by edeutsch** [comment]
I have not tested it myself, but fine to close this AFAIC and we can open a new issue if something specific is discovered.

**Comment by dkoslicki** [comment]
@saramsey Looks good to me! Note outcome of #650 may affect this issue in the future.
Very minor note @saramsey before closing, I would put `resultify` at the end of the list, as I intended the documentation to proceed roughly like the given example:
1. document create_message() first
2. document the qnode, qedge stuff
3. document the expand
4. document the overlay
5. document filter_kg
6. document filter_results
7. document resultify

Not so minor note: due to `expander` and `overlay` branch not yet being synced with demo, you're going to get some merge conflicts since I see `expander` is now missing from the `document_dsl_commands.py` `modules` list etc.

**Comment by edeutsch** [comment]
I have added documentation for create_message and add_qnode and add_qedge. It didn't work 100% right, but is pretty close. @dkoslicki said that the action= paradigm is not necessary, but the autodocumentation stuff assumes it, and working around those assumptions seems a little tricky.

Also, as a general note, create_message() isn't really explicitly necessary. If you don't reference an existing message, then it is assume that you need a new message and one is automatically created (at least in demo it is; there was a bug that I fixed that still may be present in other branches today)

**Comment by dkoslicki** [comment]
In the overlay branch, it doesn't assume the `action` is there, but branches if it is. see [these lines](https://github.com/RTXteam/RTX/blob/overlay/code/ARAX/Documentation/document_dsl_commands.py#L55-L64). I had added that specifically since `expand()` doesn't use `action=`. But LMN if it's not working in some other place

**Comment by edeutsch** [comment]
ah okay, I have a hack in there that inserts a fake action parameter to get the documentation code to run. Those fake actions can be removed once everything is merged. I'll stop on the documentation thing until everything is merged, since I guess mine is out of date.

**Comment by dkoslicki** [comment]
Yeah, I wish I could have got to the merging earlier this week, but fire-fighting before many trips on my end is under full swing. My paradigm has roughly been to merge all existing branches into mine before doing any work, just to make sure nothing gets stomped on other branches. Then add to my branch. Seems to minimize conflicts the next time someone wants to merge “my” branch into theirs.

**Comment by edeutsch** [comment]
I am uncertain if we have a clear plan. I know the plan was to merge demo into the other branches and then merge the branches down into demo. But what is not clear is who will do what by when. I think I left the Tuesday meeting thinking that others were first going going to merge demo into their branches and then at some point merge them down into demo. Was that @amykglen and @finnagin and @dkoslicki for the branches? Or were you going to do it? Or was I going to try it? If there was a clear plan of who should do what when, I have forgotten it. If there wasn't, I think we need one.

**Comment by dkoslicki** [comment]
@amykglen @finnagin @edeutsch I highly suggest that first Finn, Amy, and I merge demo into our filter, expand, and overlay branches respectively, then after resolving conflicts, Eric can merge them into demo. That way, each of us who are familiar with our own code branches can resolve conflicts there. Basically following the Github suggested SOP (as I understand it) but treating `demo` like `master`. Personally, I will also be merging filter and expand into overlay as well (to further minimize conflicts Eric might encounter). I should have a bit of time today to do that on my end

**Comment by edeutsch** [comment]
okay, great, will everyone try to get demo merged up by the end of today and report issues?

**Comment by dkoslicki** [comment]
Sure, working on overlay now

**Comment by saramsey** [comment]
OK @dkoslicki I assume I can move the `resultify` documentation to the end of the `DSL_Documentation.md` by changing the order of a list in `document_dsl_commands.py`, right?

I am happy to do that. I will make that change in the `demo` branch, this afternoon.

**Comment by dkoslicki** [comment]
@saramsey Already done! I wanted to fiddle a bit with the documentation and made this and other changes, so looks like you don't need to worry about that any more.

**Comment by dkoslicki** [comment]
@saramsey I propose we close this issue as it appears all issues and sub-issues have been addressed (and #650 has it's own issue now), which isn't really relevant to this issue any more.

**Comment by saramsey** [comment]
> @saramsey Already done! I wanted to fiddle a bit with the documentation and made this and other changes, so looks like you don't need to worry about that any more.

Thank you

**Comment by saramsey** [comment]
still need to populate `result_code` I think
